A New Proposal Process (#10)
haskellfoundation/tech-proposals: proposals/PROPOSALS.md

However, it's important to note that seeing a proposal through to its conclusion is an involved task. On the one hand, it takes time to convince people that your suggestions are a worthwhile change for hundreds of thousands of developers to accept. Particularly given the sheer volume of developers that could be affected by a proposal, its acceptance is conservative and carefully thought through. Typically, this includes many rounds of discussion with HF leaders, Haskell library maintainers, and the broader community, several iterations on the design of the proposal, and some effort at prototyping the proposed change. Often, it takes weeks to months of discussion, re-design, and prototyping for a proposal to be accepted. It is therefore important to note that seeing a proposal through to its conclusion can be time-consuming and not all proposals may end up accepted, although they may teach us all something!

If you’re motivated enough to go through this involved but rewarding process, go on with writing and keep on reading.

## What's the process for submitting a HIP?

> There are four major steps in the HIP process:
> 
> 1. Initial informal discussion (2 weeks)
> 2. Submission
> 3. Formal presentation (up to 1 month)
> 4. Formal evaluation (up to 2 months)

Reviewer: After reading the rest of this document, either the order here is wrong, or I'm misunderstanding the process.

```suggestion
There are four major steps in the HIP process:

1. Initial informal discussion (2 weeks)
2. Submission
3. Formal evaluation and revisions
4. Formal presentation after acceptance
```

Author: I think this would be most accurate: 

```
1. Initial informal discussion 
2. Submission
3. Review
4. Acceptance
5. Formal presentation
```

### Initial informal discussion (2 weeks)

Before submitting a HIP, it is required that you perform necessary preparations:

- Discuss your idea on the [Haskell.org Discourse](https://discourse.haskell.org/). Currently, we suggest cross-posting on the Haskell Foundation Slack for higher volume commentary. Create a Discourse topic under the category "Haskell Foundation", that starts with "Pre-HIP” and briefly describe what you would like to change and why you think it’s a good idea.

> - Proposing your ideas on the Discourse is not an optional step. For every change to the ecosystem, it is important to engage in Due Diligence with the community and relevant stakeholders. Use this step to promote your idea and gather early feedback on your Pre-HIP proposal. It may happen that experts and community members may have tried something similar in the past and may offer valuable advice.

Author: The optional nature of this point is questionable. Perhaps this should be an *internal* HF policy to make sure we don't tread on anyone else's work. In general I don't see a problem with people going straight to submission, though, it may cause more collisions with other people's work than we'd want.

Reviewer 2: Firstly, is there any reason why "Due Diligence" is capitalized here?

As far as the non-optionality of this, if the goal is 'eyeballs' and 'not stepping on other toes', you could mandate _either_ Discourse _or_ Slack? I think that gives some more flexibility, especially for people who prefer a more synchronous discussion medium.

Author: > Firstly, is there any reason why "Due Diligence" is capitalized here?

DD is a formal process where I'm from. Must've been a slip. 

> As far as the non-optionality of this, if the goal is 'eyeballs' and 'not stepping on other toes', you could mandate either Discourse or Slack?

If given the choice, I'd choose Discourse, since that's our stated "place to formally interface with the community". The request that the HF also do this on slack could be taken to the extreme: we could do it on Reddit, Twitter, and the mailing lists too. The point is to get the word out that we're interested by casting as wide a net as possible so as to make sure we don't overlap with any existing projects like we did with #7

Reviewer 2: The main reason I felt multiple options would be good is because not everyone is on every platform. To use myself as an example, I'm present on Slack and Twitter and pretty much nowhere else; I'm certain that there are folks who are present on one, and only one, of the options you listed there.

There is some tension here between reach, centralization and convenience; the exact point we want to hit is a little hard to determine.

Reviewer 3: For the most part, I think submission is a fine point to sort out DD. Having a public proposals process before work starts is a good opportunity to figure out what related work exists and how it interacts. In some of the cases where there was overlap, I think having had a better proposals process in place with more public discussion would have helped sort that out earlier and better.

Having pre-discussion on the discourse or whatever is however a good idea as well, since that means that there are proposals that aren't solely one person off on their own, but at least have had some initial scrutiny and feedback before they hit the requirements of a Process.

Reviewer 4: There is a real tension here: while I respect that everyone has different communication preferences, it is also important for there to be one place with the official record of what is happening. Then, an interested observer can monitor the one place. Having additional places is great -- but not at the expense of the one place. Right now, the one place the HF has designated is Discourse.

Another way to think about this is: Who should be inconvenienced more? The proposal authors or (potential) proposal readers? I think the author should bear the brunt of inconvenience: they are proposing a change that will affect potentially the entire Haskell community. This is not to be taken lightly!

As for @gbaz's point: I agree that a proposal can be a fine place to sort out DD, but I think it's not best for HF-sourced proposals, as those can appear to be in advanced stages by the time they hit this repo. See my proposed dichotomy of Proposals vs RFCs below (https://github.com/haskellfoundation/tech-proposals/pull/10#issuecomment-901184121).

Reviewer 2: @goldfirere You make an excellent point - from the point of view of inconveniencing the fewest, having a single, designated location is definitely preferable. I guess the main advantage of Discourse is 'maintained searchable history'?

Reviewer 4: While maintained searchable history is indeed fantastic, my real desire is more about an assumption of asynchronous communication. Discourse's threading model allows someone to come along several days after an announcement and see what they missed. Slack's is potentially capable of this, too, but it's not used in practice: reactions to a comment are often interleaved with new topics. Twitter similarly doesn't offer the organizational tools to be able to easily catch up if one misses a post. Discourse also has an advantage in that it's an open-source tool that we can control directly, in contrast to the other services.

Author: :100:% agree with @goldfirere here.

### Submission

> A HIP is a Markdown document written in conformance with the [process template](https://github.com/haskellfoundation/tech-proposals/blob/main/TEMPLATE.md). When such changes significantly alter an existing library or tool, the author is invited to provide a proof of concept. Delivering a basic implementation can speed up the process dramatically. If your changes are big or somewhat controversial, don’t let people hypothesize about them and show results upfront.

Reviewer 2: Is there any reason we don't _mandate_ an implementation? If the changes are _not_ big or controversial, surely the implementation can't hurt or take that much time.

Author: Some projects are proposals _to be worked on_, like #9. Throughout the course of the project, funds for branding/art needs to be decided and so on. In the case of books and other educational material, it may end up that it takes a year, which would drag the proposal process on if it were left open, and for no reason if the project is well-specified.

Reviewer 2: Yes, but none of those 'significantly alter an existing library or tool'; to me, that implies code changes. Thus, I believe mandating an implementation in those cases is fine.

Author: Ah interesting point! I'll go into detail regarding where/when an implementation is preferred

Reviewer 3: There scope of HIPs seems really broad. Prejudging ahead of time any one-size-fits-all mandates seems like a bad idea.

Author: Figuring out a release date should be discussed as part of the review, and I'd expect the Tech Track members to raise _when_ an implementation should be expected as part of the review process

A HIP is submitted as a pull request against [the official Haskell tech proposal
repo](https://github.com/haskellfoundation/tech-proposals). Within a week of receiving the pull request, members of the HF Technical Track will acknowledge your submission, validate that it conforms to the proposal template guidelines (see: [TEMPLATE.md](TEMPLATE.md)) and provide feedback to improve the overall quality of the document (if necessary). When the document conforms to the template guidelines, it is ready for formal evaluation by the HF Technical Track members, as well as the broader community. 

### Formal evaluation (up to 5 iterations)

While the majority of general technical commentary and feedback will occur in the proposals repo, it's hard to tell when a particular proposal is finalized. This what we call "Formal Evaluation". 

> Formal Evaluation of a proposal is done in iterations. The maximum number of iterations is five. These iterations take place in the HF Tech Track meetings and are usually monthly. However, they can last longer, in which case the author has more time to implement all the required changes. 

Author: HF tech track meetings are *weekly*, however, we may want to add a meeting series specifically for HIP review. The language will change as we decide what to do here.

The HF Technical Track decides the duration of the next iteration depending upon the feedback and complexity of the HIP. Consequently, authors have more time to prepare all the changes. If they finish their revision before the scheduled iteration, an HF Technical Track member will reschedule it for the next available meeting.

During every iteration, the HF Technical Track reviews the changes (updated design document, progress with the implementation, etc) to the proposal. Based on the feedback, the HIP is either:

1. **Accepted**, in which case the HF Technical Track will propose a release date.
2. **Rejected**, in which case the HIP is closed and no longer evaluated in the future.

> 3. **Postponed**, in which case the HF Technical Track sets aside the HIP under some conditions. When those conditions are met, the HIP can be resubmitted.

Author: I was planning on just doing this with github labels rather than closing the issues. What do you think?

> 3. **Postponed**, in which case the HF Technical Track sets aside the HIP under some conditions. When those conditions are met, the HIP can be resubmitted.

Reviewer: Sounds great: hat'll be a great way to prevent proposals from getting stuck in Postponed as I feared, as they would pop up when you go through open issues. Keeping issues open comes at the cost of having more issues to search through but that might be worth it. Question that arises is how do authors notify the HF TT? (To move back to 'Submitted')

> 3. **Postponed**, in which case the HF Technical Track sets aside the HIP under some conditions. When those conditions are met, the HIP can be resubmitted.

Author: > Keeping issues open comes at the cost of having more issues to search through but that might be worth it.

I think searchability of the GH issues is key here - it works well and tagging issues appropriately means label filters work similarly well. 

> Question that arises is how do authors notify the HF TT? (To move back to 'Submitted')

That's an excellent question. We should detail that. A ping to the HF Tech Track members in email/slack/discourse should be sufficient

> 3. **Postponed**, in which case the HF Technical Track sets aside the HIP under some conditions. When those conditions are met, the HIP can be resubmitted.

Author: Added a subsection detailing the use of labels and where to ping us.

4. **Under revision**, in which case the author needs to continue the formal evaluation and address all the HF Technical Track feedback. Thus, the follow-up discussion is scheduled for the next iteration.
5. **Dormant**,  in which case no changes have been made to a HIP in two iterations, it’s marked as dormant and both the PR and issue are closed. Dormant HIPs can be reopened by any person, be it the same or different authors, at which point it will start from the formal evaluation phase.

When the author and the HF Technical Track members agree on the final document, it is formally accepted: assigned a number, and scheduled for a final formal presentation.

### Formal presentation (up to 1 month)

> During the next available Technical Track meeting, the author of the HIP presents the proposal to the audience. 

Author: Good points - i'll clarify

> During the next available Technical Track meeting, the author of the HIP presents the proposal to the audience. 

Author: Okay - I've updated the language to be more detailed

### Merging the proposal

> If the HIP is accepted, the HF Technical Track will propose a release date, and the proposal will be merged directly by the HF Technical Track members.

Author: Yeah, good point. I'll strike that.

## Structure of the process

The HIP process involves the following parties:

1. The HIP Authors
2. Relevant stakeholders affected by the change

> 3. The HF Technical Track Leads

Reviewer 2: Do we have a list of these somewhere? We don't mention any _Leads_ specifically in this document.

Author: Leads should be members - thanks I missed one reference

Author: Done

### The HF Technical Track

The HF Technical Track is an experienced group of people with knowledge of the Haskell ecosystem, responsible for the strategic technical direction on behalf of the Haskell Foundation. Members are tasked with

  - communicating with the community
  - weighing in pros and cons of every proposal
  - accepting, postponing or rejecting the proposal.

HF Technical Track members should be either individuals responsible for a specific part of the Haskell ecosystem, or contributors and committers to parts of it. The members are selected by the HF CTO based on their expertise and implication in the community.

The current HF Technical Track members are:

- Emily Pillmore ([@emilypi](https://github.com/emilypi)), Haskell Foundation
- Richard Eisenberg ([@goldfirere](https://github.com/goldfirere)), Tweag
- Michael Snoyman ([@snoyberg](https://github.com/snoyberg)), FPComplete
- Andrew Lelechenko ([@Bodigrim](https://github.com/Bodigrim)), Barclays
- Davean Scies ([@davean](https://github.com/davean)), XKCD
- Edward Kmett ([@ekmett](https://github.com/ekmett)), MIRI
- Theophile Choutri ([@Kleidukos](https://github.com/Kleidukos)), Scrive
- Gil Mizrahi (@soupi)

> ### Voting
> 
> For a HIP to be accepted, it must fulfill two requirements:
> 
> - 66% of the HF Technical Track votes in favor of it
> - The HIP specifies a proposal lead (e.g. an implementor, or project director) who will represent the work done on the project.

Reviewer: This outlines the votes for moving from 'Under Revision / Under Review' to 'Accepted'. I've assumed that moving from 'Submitted' to 'Under Review / Postponed' can be done by any individual HF Technical Track member (and made a suggestion for it). My reasoning to reject as a default is that this is the only outcome (except dormant) that does not require a write-up.

Note that there is now considerably duplication between this section and the end of the 'Formal Evaluation' section. It might be better to merge this into the formal evaluation section, where it is most relevant.

```suggestion
### Voting

When a HIP is scheduled at an HF Technical Track meeting (i.e. it is in 'Under Review'), it can be held to a vote for one of the following outcomes (in case it makes a difference: the HIP will be marked according to the first on this list to have a majority):

- Accepted (needs 66% of the HF Technical Track to vote in favor, and the HIP must specify a proposal lead, e.g. an implementor or project director, from the HF Technical Track who will represent the work done on the project.)
- Dormant (needs a simple majority, an HF Technical Track member that voted in favor will close the issue and PR and mark it as Dormant)
- Postponed (needs a simple majority, an HF Technical Track member that voted to postpone will close the issue and PR, and write clear conditions for reopening)
- Revision needed (needs a simple majority. This can only be the outcome of the vote four times, for a total of five rounds. An HF Technical Track member will write up what revisions are needed.)

If none of these applies, the proposal is Rejected.
```

### Responsibilities of the members

- Review all HIPs 
- Play a role in the discussions, learn in advance about the topic if needed, and make up their mind in the voting process.

### Guests

Experts in some fields may be invited to specific meetings as guests when discussing related HIPs. Their input would be important to discuss the current state of the proposal, both its design and implementation.

## Proposal states

> The state of a proposal changes over time depending on the phase of the process and the decisions taken by the HF Technical Track . A given proposal can be in one of several states:

Author: Yes - I like that idea.

> 1. **Submitted:**  An HF Tech Track member has validated the proposal and checked that meets all the formal requirements. A pull request has been opened against the proposals repo.

Author: 👍

> 2. **Under review:** The proposal will be under review until the next available Tech Track meeting takes place.

Author: In a sense, I think we may have to add a step in between 1 and 2, which is "Under Validation", which gets the proposal up to the point where it's up for formal review. Judging by the branches here, we have 

```
Discussed 
-> Submitted
-> Under review and validated
   | looks good -> ...
   | needs work-> ...
   | postponed -> ...
   | rejected -> ...

```

That conjuction of 'validated and under review', can be split out.

3. **Under revision:** Authors are addressing the issues pinpointed throughout the discussion and feedback process from the community and/or the HF Technical Track.

6. **Dormant:** When a HIP has been under revision for more than two iterations (that is, no progress has been made since the last review), it’s considered dormant, in which case any related activity will be paralyzed and the HF Tech Track will not allocate more resources to it.

7. **Postponed:** The HIP has been postponed under some concrete conditions. When these are met, the HIP can be resubmitted.

8. **Rejected:** The HIP has been rejected with a clear and full explanation.

9. **Merged:** The HIP has been accepted and has been merged into the repo.

## How do I submit? ##

The process to submit is simple:

* Submit a Pre-HIP topic to the HF Discourse, soliciting informal feedback.
* If the feedback is favorable, fork the Haskell Foundation tech-proposals repository, [https://github.com/haskellfoundation/tech-proposals](https://github.com/haskellfoundation/tech-proposals).

> * Create a new HIP file in the `hips/` directory. Use the [HIP template](https://github.com/haskellfoundation/tech-proposals/blob/main/TEMPLATE.md)

Reviewer: Perhaps add a link back to the PROPOSALS file in the first paragraph of this template? (before the introduction)?

Author: Like a self link?

> An HFTP (*Haskell Foundation Technical Proposal*) is a process for submitting a proposal for a technical initiative to be undertaken by the Haskell Foundation (HF). This document describes how the community may propose, discuss, and implement technical initiatives to be carried out through the Haskell Foundation that will affect the Haskell ecosystem. Each HFTP goes through an iterative process in which it is discussed by the HF Technical Track (HFTT), the broader community, and any and all interested stakeholders. If a consensus is reached, the initiative is accepted and may then be executed by the Foundation (resources permitting). Only upon reaching a consensus are initiatives accepted to be merged and executed. 

Reviewer 5: What does "undertaken by the Haskell Foundation" mean? Are HF members going to execute the proposal once it is accepted? Or will the HF support the proposal's author in executing the proposal? It's important to clarify HF's role here.

1. **Concretizing ideas:** as ideas for proposals are fleshed out, so too are the resource requirements, timelines, stakeholders, and available approaches. In this sense, a proposal allows us to fully concretize a proposal idea in terms of the work needed to complete the task. 
2. **Implementation guidance:**  the myriad of subject matter experts and stakeholders affected by a proposal can comment in a centralized place to form a consensus and guide the implementation strategy for a particular task. 
3. **Resource requirements:** ideas are often discussed in the abstract, without the consideration for the cost of work required for their implementation. This process allows for discussion that reifies the resource requirements for proposals, and allows the community to more accurately weigh the cost and benefits of an implementation. 

> It's important to note that seeing a proposal through to its conclusion is an involved task. On the one hand, it takes time to convince people that your suggestions are a worthwhile change for hundreds of thousands of developers to accept. Particularly given the sheer volume of developers that could be affected by a proposal, its acceptance is conservative and carefully thought through. Typically, this includes many rounds of discussion with HF leaders, Haskell library maintainers, and the broader community, several iterations on the design of the proposal, and some effort at prototyping the proposed change. Often, it takes weeks to months of discussion, re-design, and prototyping for a proposal to be accepted. It is therefore important to note that seeing a proposal through to its conclusion can be time-consuming and not all proposals may end up accepted, although they may teach us all something!

Reviewer 5: Maybe add here how long (roughly) the proposal process take? (I think this is covered in "Formal Review" and the answer is roughly 6 months at most?)

> A HIP (*Haskell Foundation Initiative Proposal*) is a process for submitting a proposal for an initiative to be undertaken by the Haskell Foundation (HF). The main motivation of this document is to detail the primary mechanism through which the community may propose, discuss and implement initiatives to be carried out through the Haskell Foundation that will affect the Haskell ecosystem. For each HIP, all changes to the ecosystem go through a proposal process, which are is discussed by the HF Technical Track, the broader community, and any and all interested stakeholders. Only upon reaching a consensus are initiatives accepted to be merged and executed.

Reviewer 6: Maybe just "Haskell Foundation Proposal" (HFP)?  Simpler.  (Bikeshed, obviously.)

Author: I like it, but i'm sad to lose the HIP pun.

> A HIP (*Haskell Foundation Initiative Proposal*) is a process for submitting a proposal for an initiative to be undertaken by the Haskell Foundation (HF). The main motivation of this document is to detail the primary mechanism through which the community may propose, discuss and implement initiatives to be carried out through the Haskell Foundation that will affect the Haskell ecosystem. For each HIP, all changes to the ecosystem go through a proposal process, which are is discussed by the HF Technical Track, the broader community, and any and all interested stakeholders. Only upon reaching a consensus are initiatives accepted to be merged and executed.

Reviewer 6: (Wordsmithing.)  

This document describes how the community may propose....

Author: Added

> A HIP (*Haskell Foundation Initiative Proposal*) is a process for submitting a proposal for an initiative to be undertaken by the Haskell Foundation (HF). The main motivation of this document is to detail the primary mechanism through which the community may propose, discuss and implement initiatives to be carried out through the Haskell Foundation that will affect the Haskell ecosystem. For each HIP, all changes to the ecosystem go through a proposal process, which are is discussed by the HF Technical Track, the broader community, and any and all interested stakeholders. Only upon reaching a consensus are initiatives accepted to be merged and executed.

Reviewer 6: (worksmithing)

"For each HIP, all changes to the ecosystem go through a proposal process, which ..."  is hard to parse.  Maybe

Each HIP goes through an iterative process in which it is discussed by the HF Technical Track, the broader community, and any and all interested stakeholders.    If a consensus is reached, the initiative is accepted and may then be executed by the Foundation (resources permitting).

Author: Love it.

> The aim of the Haskell Improvement Process is to apply the openness and collaboration that have shaped Haskell's documentation and implementation to a process of evolving the HF and the broader Haskell ecosystem. This document captures our guidelines, commitments and expectations regarding this process.

Reviewer 7: I might be mistaken, but it seems that expansion of the HIP acronym have changed, right?
```suggestion
The aim of the HIP process is to apply the openness and collaboration that have shaped Haskell's documentation and implementation to a process of evolving the HF and the broader Haskell ecosystem. This document captures our guidelines, commitments and expectations regarding this process.
```

Author: Good catch - yes.

- Proposing your ideas on the Discourse is not an optional step. For every change to the ecosystem, it is important to engage in due diligence with the community and relevant stakeholders. Use this step to promote your idea and gather early feedback on your Pre-HIP proposal. It may happen that experts and community members may have tried something similar in the past and may offer valuable advice.

### Submission

A HIP is a Markdown document written in conformance with the [process template](https://github.com/haskellfoundation/tech-proposals/blob/main/TEMPLATE.md). When such changes significantly alter an existing library or tool, the author is invited to provide a proof of concept. Delivering a basic implementation can speed up the process dramatically. If your changes are big or somewhat controversial, don’t let people hypothesize about them and show results upfront.

> 

Author: I like that idea - it makes sense for the example you give

A HIP is submitted as a pull request against [the official Haskell tech proposal
repo](https://github.com/haskellfoundation/tech-proposals). Within a week of receiving the pull request, members of the HF Technical Track will acknowledge your submission, validate that it conforms to the proposal template guidelines (see: [TEMPLATE.md](TEMPLATE.md)) and provide feedback to improve the overall quality of the document (if necessary). When the document conforms to the template guidelines, it is ready for formal evaluation by the HF Technical Track members, as well as the broader community. 

### Formal Review (up to 5 iterations)

> While the majority of general technical commentary and feedback will occur in the proposals repo, it's hard to tell when a particular proposal is finalized. This what we call "Formal Review".  

Reviewer 6: I'm not quite sure what this paragraph means.   Here's a rephrasing that, I believe, captures your intent. But I'm not sure

The entire community is strongly encouraged to comment on, and help improve, a proposal.   Ultimately, however, the Foundation needs a mechanism to make a decision, to accept, reject, or push back a proposal.  This process is called "Formal Review", and is carried out by the HF Tech Track working group.

Reviewer 6: Given the importance of the Tech Track group, it would be good to give a link to where readers can find out what the group does, who the members are, how to become a member, etc.

Author: Good idea. I suppose these would now go hand in hand, where before we were a working group.

During every iteration, the HF Technical Track reviews the changes (updated design document, progress with the implementation, etc) to the proposal. Based on the feedback, the HIP is either:

1. **Accepted**, in which case the HF Technical Track has accepted the proposal, and it will be merged by the HF Technical Track members.
2. **Rejected**, in which case the HIP is closed and no longer evaluated in the future.
3. **Postponed**, in which case the HF Technical Track sets aside the HIP under some conditions. When those conditions are met, the HIP can be resubmitted.
4. **Under revision**, in which case the author needs to continue the formal evaluation and address all the HF Technical Track feedback. Thus, the follow-up discussion is scheduled for the next iteration. 
5. **Dormant**,  in which case no changes have been made to a HIP in two iterations, it’s marked as dormant and both the PR and issue are closed. Dormant HIPs can be reopened by any person, be it the same or different authors, at which point it will start from the formal evaluation phase.

#### Involving Stakeholders

> At this point where a propsal is being reviewed, if there are relevant stakeholders from industry or in the community (e.g. library authors and maintainers who are affected), they should be solicited for comment and brought into the discussion. Remember, proposals may sound good, but it's best to iron out as many details as possible prior to their acceptance. This means that any hesitations stakeholders have with the project should be factored into the proposal's details. 

Reviewer 4: Who is responsible for this step? Should proposal authors reach out? Or should someone from the HFTT? My take: authors should do this, and it's up to the HFTT to ensure that it's been done. Members of the HFTT can help the authors identify and reach stakeholders if that would be helpful.

Author: Agree.

> At this point where a propsal is being reviewed, if there are relevant stakeholders from industry or in the community (e.g. library authors and maintainers who are affected), they should be solicited for comment and brought into the discussion. Remember, proposals may sound good, but it's best to iron out as many details as possible prior to their acceptance. This means that any hesitations stakeholders have with the project should be factored into the proposal's details. 

Reviewer 6: Spelling "propsal"

#### Updating Proposal Statuses

> If the author of a particular proposal wants to update the status of their proposal, they should either comment on the issue and tag an HF Technical Track member with a note regarding what status they would like to update the proposal to. It is good to be redundant and raise the question in our other fora as well: Slack and Discourse. However, Github should be the first choice. 

Reviewer 4: "tag an HFTT member"

Which one? I'm trying to think of myself as an author, and this instruction seems confusing, requiring me to look up the HFTT membership, track down their GitHub names, and then arbitrarily choose one, hoping not to offend anyone. We should either designate an individual (essentially: the HFTT secretary) or have some sort of group tag (does GitHub even have this?).

"GitHub should be the first choice": This, to me, implies that an author who dislikes GitHub may choose to just make a small comment on Slack. That comment might get missed, and then the author gets frustrated. I think we should clarify that all official requests for action on a proposal take place through GitHub.

Author: Github does have group tags (ala @haskellfoundation/tech-proposals), which requires adding HFTT members to a team on this repo. This is easy to do, and I would prefer this as opposed to adding a new role.

### Formal presentation 

> After a proposal is merged, the author will present KPI's and deliverables that define the proposal to the HF Technical track. During the next available Technical Track meeting after acceptance, the author or designated lead (as defined by the proposal) should give a short 5-10min presentation consisting of the following: 

Reviewer 4: What is a KPI? I really don't know.

This step seems surprising to me. During the whole deliberation process, it seems that the author is not generally invited to present their case. And then after the proposal is accepted, they can? I'm not sure I see the logic here.

Reviewer 6: I too am a bit surprised.  I'm inclined to suggest that this document ends with acceptance. Then there is some separate process by which the HF board, led by its ED and CTO have to prioritise resources, choose initiatives, etc.   That may or may not involve project presentations, but it'll involve a lot else.  But it's beyond the scope of this document. 

This links with my remarks below about the distinction between acceptance and execution.

Author: > What is a KPI

Key Performance Indicators - it's a term used in project management in Agile settings. I'll just spell it out. 

> Suprising

The logic here is that the accepted proposal represents the finalized proposal state, while the presentation gives the authors a chance to present the finalized proposal to the track as a kind of "first standup". The HFTT gets to meet the representative who will be giving future updates, as well as give the broader HFTT (who may all not have seen the proposal) the chance to see the finished product. Of course there are questions here: is this necessary? Perhaps not. I may have gone beyond the scope here.

Reviewer 4: Yes, I tend to think that this is beyond the process. That is, once we agree to pursue a project, the proposals process is complete. It's quite reasonable for the HFTT to both oversee the proposal process and the implementation process, but the meeting described here sounds like part of the implementation process. In my opinion, there's not as much of a need to codify implementation processes, as there will be big differences between e.g. volunteer work and work that the HF pays for.

However, there is still an open question: is there an opportunity for proposal authors to make their case live to the HFTT? Maybe this step is optional, but if so, who decides whether authors get to present? Does the HFTT invite them based on their interest? Does the author decide they wish to attend in person? I tend to think it's fairest if we give proposers the opportunity to make their case to us -- all the more so because I suspect that HFTT members will sometimes also be authors, and naturally HFTT members will advocate for their proposals in meetings.

> After a proposal is merged, the author will present KPI's and deliverables that define the proposal to the HF Technical track. During the next available Technical Track meeting after acceptance, the author or designated lead (as defined by the proposal) should give a short 5-10min presentation consisting of the following: 

Reviewer 6: I think it would be good to state explicitly that the HF *accepting* a proposal does not imply that the HF commits to *executing* that proposal.  The HF only has limited resources -- and it complicates technical discussion too much if it's tangled up with resource considerations.  

So it's helpful to distinguish between (a) "do we agree that it would be good to do X?" from (b) "Is X (a good thing) the most important thing to execute on next, in competition with Y and Z?".

Of course, if the HF has helped to evolve a consensus that X is a good thing, it may well be that, thus encouraged, someone in the Haskell community steps up to do it.   This happens all the time with GHC proposals.

I think it would be helpful to make this explicit, somewhere early in this document

1. **High level summary:** the author will present the high-level abstract for what the proposal intends to accomplish. 
2. **Timeline:** In what timeline, and should outline possible hurdles and risks that affect the timeline. 
3. **Resources:** What resources are required in order to take the project from start to completion.
4.  **Contact:** Who is going to update the HF Technical Track members, and on what schedule.

> Once the formal presentation is given, the HF Technical Track members will track the progress over time at our ongoing meetings. This also gives the track members time to introduce themselves and and we can all get to know each other. Currently, HF Technical Track meetings are held weekly on Tuesdays at 11:30am-12:30pm UTC-5 (EST). 

Reviewer 4: "This also gives the track members time to introduce themselves and and we can all get to know each other." I don't see how this sentence relates to the rest of the paragraph.

It's true that current meetings are 11:30-12:30 EST, but that's a bit confusing: no place on earth is in the EST time zone at the moment (we're in EDT for summer). Will this timeslot stay the same after daylight saving time ends? Will we remember to update this when changing the schedule? Instead, I think we should have an HFTT home page of sorts (maybe the README to this repo? Maybe haskell.foundation/projects?) that lists the current status of the HFTT, including membership, meeting times, and how to join. Then, we just refer to that page from here.

> 

Reviewer 4: I've been expecting to see a notion of shepherd here somewhere. That is, for each proposal, there should be an individual who feels responsible for keeping it in line. Maybe this is always you, @emilypi, for every proposal. But I think having responsible individuals, instead of a responsible committee, is more likely to lead to forward action.

## Structure of the process

> 

Reviewer 6: This is a bit of an odd section.  What would be lost if we deleted it?

Author: I wouldn't want to scrap it completely, because it tells precisely how voting works, who will be voting, and so on. I'm fine with losing the preamble and the first section (The HIP authors). Those don't bring much.

Author: Agreed - removed in latest commits

The HIP process involves the following parties:

1. The HIP Authors
2. Relevant stakeholders affected by the change
3. The HF Technical Track members

### The HIP Authors

> Authors are responsible for building consensus within the community and documenting dissenting opinions before the HIP is officially discussed by the HIP HF Technical Track . Their goal is to convince the HF Technical Track that their proposal is useful and addresses pertinent problems in the Haskell ecosystem as well as interactions with already existing features. Authors can change over the life-cycle of the HIP.

Reviewer 4: ```suggestion
Authors are responsible for building consensus within the community and documenting dissenting opinions before the HIP is officially discussed by the HIP HF Technical Track. Their goal is to convince the HF Technical Track that their proposal is useful and addresses pertinent problems in the Haskell ecosystem as well as interactions with already existing features. Authors can change over the life-cycle of the HIP.
```

> Authors are responsible for building consensus within the community and documenting dissenting opinions before the HIP is officially discussed by the HIP HF Technical Track . Their goal is to convince the HF Technical Track that their proposal is useful and addresses pertinent problems in the Haskell ecosystem as well as interactions with already existing features. Authors can change over the life-cycle of the HIP.

Reviewer 6: If the decision making body is the Tech Track working group, then I assume HIPs are intended to be primarily technical.   What about proposals that are more social than technical, e.g. "We should run an annual conference" or "HF should affiliate to group X"?

Rather than try to be comprehensive, I suggest restricting this process to stuff within the scope of the Tech Track task force -- and try to write down what that scope is.  (In a word "technical".)

> ### The HF Technical Track

Reviewer 6: Don't we need a different document that describes, for the Tech Track working group
* Purpose
* Membership
* How to become a member, membership terms etc

etc?  Does that really belong here?   If reviewing HIPs is the sole purpose of the Tech Track wg then maybe yes it belongs here (in which case we need to say more, so it's a complete description of the working group).  But I suspect it has some broader roles and deserves its own document.

The HF Technical Track is an experienced group of people with knowledge of the Haskell ecosystem, responsible for the strategic technical direction on behalf of the Haskell Foundation. Members are tasked with

  - communicating with the community
  - weighing in pros and cons of every proposal
  - accepting, postponing or rejecting the proposal.

> HF Technical Track members should be either individuals responsible for a specific part of the Haskell ecosystem, or contributors and committers to parts of it. The members are selected by the HF CTO based on their expertise and implication in the community.

Reviewer 4: Did you mean "implication"? Perhaps "reputation" would be better?

I may have complained about this previously, but I'm now OK with you, @emilypi, unilaterally selecting the HFTT. We serve at your pleasure. But I do think there should be a process by which a community member can apply to join. Having open applications works against a situation where you choose only people close to you -- not because of bias, but because those are the people you're aware of.

Is this list of people up-to-date? Some names here are not regular attendees of our meetings and may be surprised to be listed. Best to double-check with everyone. (I explicitly agree to serve.)

Author: > Did you mean "implication"? Perhaps "reputation" would be better?

In a sense, implication is both the correct word and too strong a word here. We don't want the HFTT to be composed exclusively of Haskell titans, which sets the bar extremely high and gives the wrong impression. I'd be perfectly fine with having enthusiastically dedicated students, as well as representatives from industry, in addition to those big Haskell names. But implication sounds a bit... political. Here, we'd want to say something broader, like taking a sample of the community.

> The HFTT is an experienced group of people with knowledge of the Haskell ecosystem, responsible for the strategic technical direction on behalf of the Haskell Foundation. Members are tasked with

Author: Fixed in latest commits

The current HF Technical Track members are:

- Emily Pillmore ([@emilypi](https://github.com/emilypi)), Haskell Foundation
- Richard Eisenberg ([@goldfirere](https://github.com/goldfirere)), Tweag
- Michael Snoyman ([@snoyberg](https://github.com/snoyberg)), FPComplete
- Andrew Lelechenko ([@Bodigrim](https://github.com/Bodigrim)), Barclays
- Davean Scies ([@davean](https://github.com/davean)), XKCD
- Edward Kmett ([@ekmett](https://github.com/ekmett)), MIRI
- Theophile Choutri ([@Kleidukos](https://github.com/Kleidukos)), Scrive
- Gil Mizrahi (@soupi)

> ### Voting

Reviewer 4: If you, @emilypi, unilaterally select the HFTT, then having these rules around voting seem unenforceable: you could always kick someone off the HFTT just during a vote and then reinstate them. So I think it simplifies matters considerably to make the HFTT a labor-producing and advisory body. We can vote, but our vote is a recommendation to you, who makes the decisions.

Author: > make the HFTT a labor-producing and advisory body.

I think this idea is in line with what we want as a foundation.

- Accepted (needs 66% of the HF Technical Track to vote in favor, and the HIP must specify a proposal lead, e.g. an implementor or project director, from the HF Technical Track who will represent the work done on the project.)
- Dormant (needs a simple majority, an HF Technical Track member that voted in favor will close the issue and PR and mark it as Dormant)
- Postponed (needs a simple majority, an HF Technical Track member that voted to postpone will close the issue and PR, and write clear conditions for reopening)
- Revision needed (needs a simple majority. This can only be the outcome of the vote four times, for a total of five rounds. An HF Technical Track member will write up what revisions are needed.)

If none of these applies, the proposal is Rejected.

### Responsibilities of the members

- Review all HIPs 
- Play a role in the discussions, learn in advance about the topic if needed, and make up their mind in the voting process.

### Guests

Experts in some fields may be invited to specific meetings as guests when discussing related HIPs. Their input would be important to discuss the current state of the proposal, both its design and implementation.

> ## Proposal states

Reviewer 6: How can I see the state of my proposal?  How can I see the state of all proposals?  Maybe no need to answer here, but once in operation these should be easy questions to answer.

Author: Good point. I'll add a sentence about using the Github Label system

> The state of a proposal changes over time depending on the phase of the process and the decisions taken by the HF Technical Track. A given proposal can be in one of several states:

Reviewer 6: A flow diagram would be quite useful here.

Author: Agreed

6. **Postponed:** The HIP has been postponed under some concrete conditions. When these are met, the HIP can be resubmitted.
7. **Rejected:** The HIP has been rejected with a clear and full explanation.
8. **Merged:** The HIP has been accepted and has been merged into the repo.

## How do I submit? ##

The process to submit is simple:

* Submit a Pre-HIP topic to the HF Discourse, soliciting informal feedback.
* If the feedback is favorable, fork the Haskell Foundation tech-proposals repository, [https://github.com/haskellfoundation/tech-proposals](https://github.com/haskellfoundation/tech-proposals).
* Create a new HIP file in the `hips/` directory. Use the [HIP template](https://github.com/haskellfoundation/tech-proposals/blob/main/TEMPLATE.md)
 * Make sure the new file follows the format: `YYYY-MM-dd-{title}.md`. Use the current date for `YYYY-MM-dd`.
 * Use the [Markdown Syntax](http://daringfireball.net/projects/markdown/syntax) to write your HIP.
* Commit your changes to your forked repository
* Create a new [pull request](https://github.com/haskellfoundation/tech-proposals/pull/new).

> * Notify the Haskell HIP team on the Slack instance or on Discourse (or both!).

Reviewer 4: Is this last step necessary? I would expect GitHub to do this notification.

Reviewer: I agree that GitHub should, but I also really like this step. It's a fallback in case someone erroneously forked a fork of this repo and sent the pull request there instead. It also establishes an extra line of communication to the author. You may not know their slack/discourse/email name otherwise. Good for comments like 'awesome proposal', or 'I got the impression that what you were discussing in Discourse was a different proposal?' or 'I don't think I understand any of this small-step semantics stuff, we may need to wait for goldfirere to get back from his trip so we can make sense of it'.

A HIP (*Haskell Improvement Process*) is a process for submitting changes to the Haskell language ecosystem. Its main motivation is to become the primary mechanism to propose, discuss and implement ecosystem changes through the Haskell Foundation. In this process, all changes to the ecosystem go through a proposal process, called Haskell Improvement Proposals (HIPs), which are openly discussed by the Haskell Foundation Technical Trackm and any and all relevant or interested stakeholders. Only upon reaching a consensus are accepted to be merged and executed.

The aim of the Haskell Improvement Process is to apply the openness and collaboration that have shaped Haskell's documentation and implementation to the process of evolving the language and its ecosystem. This document captures our guidelines, commitments and expectations regarding this process.

> ## Why Write a Proposal?

Author: Only mild changes were made to the language

If you’re motivated enough to go through this involved but rewarding process, go on with writing and keep on reading.

## What's the process for submitting a HIP?

There are four major steps in the HIP process:

1. Initial informal discussion (2 weeks)
2. Submission
3. Formal presentation (up to 1 month)
4. Formal evaluation (up to 2 months)

### Initial informal discussion (2 weeks)

Before submitting a HIP, it is required that you perform necessary preparations:

- Discuss your idea on the [Haskell.org Discourse](https://discourse.haskell.org/)  currently, we suggest cross-posting on the Haskell Foundation Slack for higher volume commentary) Create a Haskell Foundation topic that starts with “Pre-HIP” and briefly describe what you would like to change and why you think it’s a good idea.

> - Proposing your ideas on the Discourse is not an optional step. For every change to the ecosystem, it is important to *publicly* gauge interest from the community and relevant stakeholders. Use this step to promote your idea and gather early feedback on your informal proposal. It may happen that experts and community members may have tried something similar in the past and may offer valuable advice.

Author: From @goldfirere: 

> But the proposal process itself is public, no?

Author: From @myShoggoth 

> can

(re: "may")

Author: This one is tricky, because it's both a stopgap against people not doing their due diligence, but also to make sure Pre-HIPs are circulated as widely as is reasonable possible. However, it *is* a slightly optional step in my mind: the proposal process does indeed handle the "public" part of this. Perhaps this should be an *internal* policy, to make sure that anything we propose goes through this informal due diligence to make sure we're not stepping on anyone's toes. It may be worth specifying additional criteria for how proposals arise from within the HF.

> Within two weeks after your submission of the pre-HIP to the Discourse, a Technical Track Lead will intervene and advise you whether your idea can be submitted as a HIP or needs more work.

Author: From  @myShoggoth 

> "of"

re: "after"

Author: From @goldfirere: 

> Who's that? You list the Tech Track members below. Is a Lead different?

Author: and 

> This feels slightly like censoring. We must give permission in order to even write a proposal?

Author: Re: censoring, no. I should strike that. There's no real consequence here if there's been at least some DD for the proposal.

### Submission

After receiving the green light from a Technical Track Lead, you can write up your idea and submit it as a HIP.

A HIP is a Markdown document written in conformance with the [process template](https://github.com/haskellfoundation/tech-proposals/blob/main/TEMPLATE.md). When such changes significantly alter an existing library or tool, the author is invited to provide a proof of concept. Delivering a basic implementation can speed up the process dramatically. If your changes are big or somewhat controversial, don’t let people hypothesize about them and show results upfront.

A HIP is submitted as a pull request against [the official Haskell tech proposal

> repo](https://github.com/haskellfoundation/tech-proposals). Within a week of receiving the pull request, the HF Technical Track Lead will acknowledge your submission, validate it and engage into some discussions with the author to improve the overall quality of the document (if necessary).

Author: From @goldfirere: 

> What does "validate" mean here?

> When you and the HF Technical Track Lead agree on the final document, it is formally accepted for review: assigned a reviewer and scheduled for formal presentation.

Author: From @goldfirere 

> Where is there chance for detailed community feedback? Authors get light community feedback from the pre-proposal phase, but then it sounds like all tasks are held within the HF. Is that intended?

### Formal presentation (up to 1 month)

> During the next available Technical Track meeting, the appointed reviewer presents the HIP to the audience and kick starts the initial discussion.

Author: From @goldfirere: 

> This sounds like the author is not available to present their own work. I worry that aspects of the proposal could very easily get misrepresented this way.

If the leads agree that following through the HIP is a good idea, then the following happens:

1. The HIP is assigned a number.

> 2. The HIP pull request is merged into the official repo, and the merged document becomes the official webpage of the proposal.

Author: From @goldfirere 

> This surprises me. So a proposal is merged before the proposal is accepted?

3. An issue to discuss the HIP is opened at the official repo. Then, the reviewer submits the initial feedback from the HF Technical Track.

> 4. An implementation is requested (if not already present).

Author: From @goldfirere: 

> An implementation is requested before the proposal is accepted (and indeed, subject to change)? That seems backwards to me.

> Otherwise, the HIP is rejected. The reviewer submits the collected feedback as a comment to the pull request, and the pull request is closed.

Author: From @goldfirere: 

> Is there a possibility for revision?

### Formal evaluation (up to 5 iterations)

> Evaluation of a proposal is done in iterations. The maximum number of iterations is five. These iterations take place in the HIP meetings and are usually monthly. However, they can last longer, in which case the author has more time to implement all the required changes.

Author: From @goldfirere 

> What are HIP meetings? Are these the same as tech track meetings?

> The HF Technical Track decides the duration of the next iteration depending upon  the feedback and complexity of the HIP. Consequently, authors have more time to prepare all the changes. If they finish their revision before the scheduled iteration, the Process Lead will reschedule it for the next available meeting.

Author: From @goldfirere: 

> This is a new term. Who is the Process Lead?

> During every iteration, the appointed reviewer presents the changes (updated design document, progress with the implementation, etc) to the HF Technical Track. Based on the feedback, the HIP is either:

Author: From @goldfirere:

> This sounds contradictory, so I must be misunderstanding something. I thought an iteration included a revision. So this cannot happen? Or is "iteration" just a synonym for "month"?

1. **Accepted**, in which case the HF Technical Track will propose a release date.
2. **Rejected**, in which case the HIP is closed and no longer evaluated in the future.
3. **Postponed**, in which case the HF Technical Track sets aside the HIP under some conditions. When those conditions are met, the HIP can be resubmitted.
4. **Under revision**, in which case the author needs to continue the formal evaluation and address all the HF Technical Track feedback. Thus, the follow-up discussion is scheduled for the next iteration.

If no changes have been made to a HIP in two iterations, it’s marked as dormant and both the PR and issue are closed. Dormant HIPs can be reopened by any person, be it the same or different authors, at which point it will start from the formal evaluation phase.

> ### Merging the proposal

Author: From @goldfirere 

> But I thought the proposal was already merged, as described above.

If the HIP is accepted, the HF Technical Track will propose a release date, and the proposal will be merged directly by the Track Leads.

## Structure of the process

The HIP process involves the following parties:

1. The HIP Authors

> 2. Relevant stakeholders affected by the change

Author: From @goldfirere: 

> Yes! But where do these people have a role in this process? I don't see it.

Author: I've added the relevant language in the Formal Review section

3. The HF Technical Track Leads

### The HIP Authors

> Authors are responsible for building consensus within the community and documenting dissenting opinions before the HIP is officially discussed by the HIP HF Technical Track . Their goal is to convince the HF Technical Track that their proposal is useful and addresses pertinent problems in the language as well as interactions with already existing features. Authors can change over the lifecycle of the HIP.

Author: From @goldfirere: 

> How should this happen? The process doesn't really allow for this step.

Author: and 

> Is this a sub-track of the usual Technical Track? Presumably not, but it sounds like it is here.

### The HF Technical Track

The HIP HF Technical Track is an experienced group of people with knowledge of the Haskell ecosystem, responsible for the strategic technical direction on behalf of the Haskell Foundation. Members are tasked with

  - communicating with the community
  - weighing in pros and cons of every proposal
  - accepting, postponing or rejecting the proposal.

> HF Technical Track members should be either individuals responsible for a specific part of the Haskell ecosystem, or contributors and committers to parts of it. Exceptionally, members may also be important representatives of the community with a high technical knowledge to understand the implications of every proposal and participate into the discussions. The members are elected by the HF CTO based on their expertise and implication in the community.

Author: From @goldfirere: 

> I don't see how this "exception" is different from the previous sentence.

Author: @myShoggoth: 

> selected?

re: elected

Author: From @goldfirere: 

> I agree that "elected" seems like the wrong word here. I would say "appointed". But how does one offer themselves for consideration here? This seems very closed-door. I'm not comfortable with this.

The current HF Technical Track members are:

- Emily Pillmore ([@emilypi](https://github.com/emilypi)), Haskell Foundation

> - Andrew Boardman ([@myShoggoth](https://github.com/myShoggoth)), Haskell Foundation

Author: From @myShoggoth: 

> Remove

- Richard Eisenberg ([@goldfirere](https://github.com/goldfirere)), Tweag
- Michael Snoyman ([@snoyberg](https://github.com/snoyberg)), FPComplete
- Chris Dornan ([@cdornan](https://github.com/cdornan)), IRIS Connect
- Andrew Lelechenko ([@Bodigrim](https://github.com/Bodigrim)), Barclays
- Davean Scies ([@davean](https://github.com/davean)), XKCD
- Edward Kmett ([@ekmett](https://github.com/ekmett)), MIRI
- Theophile Choutri ([@Kleidukos](https://github.com/Kleidukos)), Scrive

### Reviewers

The HF Technical Track assigns every proposal to one of its members, who becomes the reviewer. The main tasks of the reviewer are the following:

1. Discuss unclear points with the authors,
2. Help them address their issues and questions,
3. Provide them feedback from the discussions in the meetings, and
4. Explain the latest progress in every meeting.

### Voting

For a HIP to be accepted, it must fulfill two requirements:

- 66% of the HF Technical Track votes in favor of it

> - The HIP specifies a lead who will represent the work done on the project.

Author: From @goldfirere: 

> Is this a different person that the lead mentioned elsewhere? Presumably. Let's use a different word then.

### Responsibilities of the members

> - Review the proposals they are assigned to. The HF Technical Track will always notify them two weeks in advance, at minimum.

Author: From @goldfirere: 

> Who, precisely? The HF Technical Track, as an entity, does not have the capability of producing emails.

- Play a role in the discussions, learn in advance about the topic if needed, and make up their mind in the voting process.

### Guests

> Experts in some fields may be invited to concrete meetings as guests when discussing related HIPs. Their input would be important to discuss the current state of the proposal, both its design and implementation.

Author: From @myShoggoth: 

> specific or particular

(re: concrete)

## Proposal states

The state of a proposal changes over time depending on the phase of the process and the decisions taken by the HF Technical Track . A given proposal can be in one of several states:

1. **Validated:** The Process Lead has validated the proposal and checked that meets all the formal requirements.

> 2. **Numbered:** The HF Technical Track agrees that the proposal is a valid document and it’s worth considering it. Then, the Process Lead gives it a number.

Author: From @goldfirere: 

> What's the difference between Validated and Numbered?

3. **Awaiting review:** The proposal has been scheduled to be reviewed for a concrete date.

> 4. **Under review:** Once the author has delivered a new version, the proposal will be under review until the next available HIP meeting takes place.

Author: From @goldfirere: 

> This sounds identical to the previous state.

5. **Under revision:** Authors are addressing the issues pinpointed by the HF Technical Track or working on the implementation.
6. **Dormant:** When a HIP has been under revision for more than two iterations (that is, no progress has been made since the last review), it’s considered dormant, in which case any related activity will be paralysed and the Process Lead will not allocate more resources to it.
7. **Postponed:** The HIP has been postponed under some concrete conditions. When these are met, the HIP can be resubmitted.
8. **Rejected:** The HIP has been rejected with a clear and full explanation.

> 9. **Accepted:** The HIP has been accepted and it’s waiting for the merge into the repo.

Author: From @goldfirere: 

> This state sounds very transitory... as in, it will last only a few minutes, right?

## How do I submit? ##

> The process to submit is simple:

Author: From @goldfirere: 

> What about the pre-discussion phase? That seems missing here.

Author: (This was added as the Pre-HIP topic blurb)

* Fork the Haskell Foundation tech-proposals repository, [https://github.com/haskellfoundation/tech-proposals](https://github.com/haskellfoundation/tech-proposals).
* Create a new HIP file in the `hips/pending/` directory. Use the [HIP template](https://github.com/haskellfoundation/tech-proposals/blob/main/TEMPLATE.md)
 * Make sure the new file follows the format: `YYYY-MM-dd-{title}.md`. Use the proposal date for `YYYY-MM-dd`.
 * Use the [Markdown Syntax](http://daringfireball.net/projects/markdown/syntax) to write your HIP.
* Commit your changes to your forked repository

> * Create a new [pull request](https://github.com/haskellfoundation/tech-proposals/pull/new). Notify the Haskell HIP team on the Slack instance.

Author: From @goldfirere: 

> Is this really necessary? The HIP team (who is this? presumably the tech track) should probably just get notifications from GH directly.
